source/standards/test-driven-development.html.md.erb

```diff
@@ -0,0 +1,48 @@
+---
+title: Test-driven development (TDD)
+last_reviewed_on: 2024-10-21
+review_in: 6 months
+---
+# <%= current_page.data.title %>
+
+GDS advocates for agile software development practices because we believe that flexibility and responsiveness-to-change are key to building software that meets the needs of our users.
+
+Test-driven development (TDD) is a core agile software development practice which aims to build flexibility and correctness into production software.
+
+The process starts with a set of desired behaviours for a piece of code. This list could include happy-path behaviours, error cases, and interaction points with other components.
+
+The inner-loop of TDD is also commonly referred to as "Red-Green-Refactor"; red to represent a failing test, green to represent all passing tests, and refactoring only when all tests are passing.
+
+<%= partial 'partials/svg-tdd-red-green-refactor' %>
+
+### Red - write a failing test
+Turn exactly one item on the list into an actual, concrete, runnable test. This test should fail when you run it, and it should fail in the way you expect. If it fails unexpectedly, that could indicate a functionality gap or a misimplementation. For example, setting a field as immutable when you expect to mutate it on later interactions.
+
+### Green - make all tests pass
+Change the code to make all tests, including the new test, pass. If you discover that you are missing a behaviour or test, add it to the list. Tests should pass in a timely manner and slow-running tests should be investigated in the refactoring stage. You may choose to reduce the size of your test case or decompose the functionality into multiple, separately testable, components.
+
+### Refactor - iterate on your design
+
+Improve the design and implementation of the code by removing duplication, utilising well known patterns and structures. You should let yourself be guided by the tests you've written; if a test is hard to write or you find yourself having to update many tests to support a new test without using automated refactoring tools, then take a step back and examine why.
+
+These steps are repeated until your set of behaviours is exhausted.
+
+## Why we recommend it
+
+The advantage of writing the tests before the implementation is that the behaviour of the code you will write is defined first, rather than implementing the change and retroactively applying tests.
```

Comment:
Do you think it's also worth mentioning TDD as an aid to thinking here?

Personally I find TDD forces me to confront a bunch of design questions with a degree of clarity that I otherwise miss. It's so easy to charge ahead on a problem with a fairly fuzzy idea in mind. TDD helps turn fuzzy dreams into firm plans.

Reply:
I've put something about software design in the "refactor" section above.